compiler/rustc_codegen_ssa/src/base.rs

```diff
@@ -842,17 +838,11 @@ impl CrateInfo {
             if tcx.is_no_builtins(cnum) {
                 info.is_no_builtins.insert(cnum);
             }
-            let missing = tcx.missing_lang_items(cnum);
-            for &item in missing.iter() {
-                if let Ok(id) = lang_items.require(item) {
-                    info.lang_item_to_crate.insert(item, id.krate);
-                }
-            }
 
             // No need to look for lang items that don't actually need to exist.
-            let missing =
-                missing.iter().cloned().filter(|&l| lang_items::required(tcx, l)).collect();
-            info.missing_lang_items.insert(cnum, missing);
+            info.missing_lang_items.extend(
+                tcx.missing_lang_items(cnum).iter().filter(|l| lang_items::required(tcx, **l)),
+            );
```

Comment:
This needs to be skipped if the current crate is not in `used_crates`.

Reply:
Seems reasonable, but `compiler\rustc_passes\src\weak_lang_items.rs` and friends do not currently contain any similar logic.

Proc macro crates have empty missing lang item sets by definition (due to https://github.com/rust-lang/rust/pull/100404#discussion_r946702199).
And as for `#[no_link]` crates, I'm pretty sure nobody ever tried to test a configuration where language items are put into them.
Not sure about rlibs used from proc macros. Anyway, this condition won't hurt, just may uncover some preexisting issues in corner cases.

Reply:
> Not sure about rlibs used from proc macros.

AFAIK those are included in the crates list, but not as used crates.